homeassistant/components/sonos/switch.py

```diff
@@ -0,0 +1,224 @@
+"""Entity representing a Sonos battery level."""
+from __future__ import annotations
+
+import datetime
+import logging
+
+from pysonos.alarms import Alarm
+from pysonos.exceptions import SoCoUPnPException
+
+from homeassistant.components.switch import ENTITY_ID_FORMAT, SwitchEntity
+from homeassistant.const import ATTR_TIME
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
+from . import SonosData
+from .const import (
+    DATA_SONOS,
+    DOMAIN as SONOS_DOMAIN,
+    SONOS_ALARM_UPDATE,
+    SONOS_CREATE_ALARM,
+)
+from .entity import SonosEntity
+from .speaker import SonosSpeaker
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_DURATION = "duration"
+ATTR_ID = "alarm_id"
+ATTR_PLAY_MODE = "play_mode"
+ATTR_RECURRENCE = "recurrence"
+ATTR_SCHEDULED_TODAY = "scheduled_today"
+ATTR_VOLUME = "volume"
+ATTR_INCLUDE_LINKED_ZONES = "include_linked_zones"
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Sonos from a config entry."""
+
+    async def _async_create_entity(speaker: SonosSpeaker) -> None:
+        data_sonos = hass.data[DATA_SONOS]
+
+        for alarm in speaker.available_alarms:
+            if alarm.alarm_id not in data_sonos.alarms:
+                _LOGGER.debug("Creating alarm with id %s", alarm.alarm_id)
+                entity = SonosAlarmEntity(alarm, data_sonos, speaker)
+                async_add_entities([entity])
+                data_sonos.alarms.append(alarm.alarm_id)
+                config_entry.async_on_unload(
+                    async_dispatcher_connect(
+                        hass, SONOS_ALARM_UPDATE, entity.async_update
+                    )
+                )
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(hass, SONOS_CREATE_ALARM, _async_create_entity)
+    )
+
+
+def get_speaker_from_uid(uid: str, data_sonos: SonosData) -> SonosSpeaker:
+    """Get speaker from player uid."""
+    for discovered_uid in data_sonos.discovered.keys():
+        if discovered_uid == uid:
+            return data_sonos.discovered[discovered_uid]
+
+    raise RuntimeError("No configured Sonos speaker has been found to match the alarm.")
```

Comment:
`discovered` is a dict keyed by `uid`. Is this method necessary?

Reply:
Oh 😮 
I totally agree!